lib/puppet/parser/functions/new.rb

```diff
@@ -0,0 +1,530 @@
+Puppet::Parser::Functions::newfunction(
+  :new,
+  :type => :rvalue,
+  :arity => -1,
+  :doc => <<-DOC
+Creates a new instance/object of a given data type.
+
+This functions makes it possible to create new instances of
+concrete data types. If a block is given it is called with the
+just created instance as an argument.
+
+Calling this function is equivalent to directly
+calling the data type:
+
+**Example:** `new` and calling type directly are equivalent
+
+```puppet
+$a = Integer.new("42")
+$b = Integer("42")
+```
+
+These would both convert the string `"42"` to the decimal value `42`.
+
+**Example:** arguments by position or by name
+
+```puppet
+$a = Integer.new("42", 8)
+$b = Integer({from => "42", radix => 8})
+```
+
+This would convert the octal (radix 8) number `"42"` in string form
+to the decimal value `34`.
+
+The new function supports two ways of giving the arguments:
+
+* by name (using a hash with property to value mapping)
+* by position (as regular arguments)
+
+Note that it is not possible to create new instances of
+some abstract data types (for example `Variant`). The data type `Optional[T]` is an
+exception as it will create an instance of `T` or `undef` if the
+value to convert is `undef`.
+
+The arguments that can be given is determined by the data type.
+
+> An assertion is always made that the produced value complies with the given type constraints.
+
+**Example:** data type constraints are checked
+
+```puppet
+Integer[0].new("-100")
+```
+
+Would fail with an assertion error (since value is less than 0).
+
+The following sections show the arguments and conversion rules
+per data type built into the Puppet Type System.
+
+Conversion to Optional[T] and NotUndef[T]
+-----------------------------------------
+
+Conversion to these data types is the same as a conversion to the type argument `T`.
+In the case of `Optional[T]` it is accepted that the argument to convert may be `undef`.
+It is however not acceptable to give other arguments (than `undef`) that cannot be
+converted to `T`.
+
+Conversion to Integer
+---------------------
+
+A new `Integer` can be created from `Integer`, `Float`, `Boolean`, and `String` values.
+For conversion from `String` it is possible to specify the radix (base).
+
+```puppet
+type Radix = Variant[Default, Integer[2,2], Integer[8,8], Integer[10,10], Integer[16,16]]
+
+function Integer.new(
+  String $value,
+  Radix $radix = 10
+)
+
+function Integer.new(
+  Variant[Numeric, Boolean] $value
+)
+```
+
+* When converting from `String` the default radix is 10
```

Comment:
Each line in this bulleted list should end in `.`, including the ones currently ending with commas.